File: core/HaskellAnywhere.java
package core;

import result.CommunicationResult;

public class HaskellAnywhere {
	
	private String program;
	private String mainFunction;
	private final String defaultMainFunction = "main = print $ ";
	private RexTester compiler;
	
	public HaskellAnywhere() {
		this.program = "";
		this.mainFunction = "";
		this.compiler = new RexTester();
	}
	
	public CommunicationResult input(String arg0) {
		CommunicationResult result;
		StringAnalyzer analyzer = new StringAnalyzer(arg0);
		if (analyzer.isMain()) {
			mainFunction = this.defaultMainFunction;
			mainFunction += analyzer.getText();
			result = submit();
		} else {
			program += analyzer.getText() + "\n";
			result = ResultFactory.createAddDefineResult("{\"AddedDefine\":\"" + analyzer.getText() + "\"}");
		}
		return result;
	}
	
	public CommunicationResult submit() {
		return this.compiler.post(program + mainFunction, "");
	}
	
	public String getProgram() {
		return this.program;
	}
	
	public String getMainFunction() {
		return this.mainFunction;
	}
}

File: src/com/lucciola/haskell/HaskellAnywhere.kt
package com.lucciola.haskell

import com.lucciola.result.CommunicationResult

class HaskellAnywhere {

    var program: String = ""
        private set
    var mainFunction: String = ""
        private set
    private val defaultMainFunction = "main = print $ "
    private val compiler: RexTester = RexTester()

    fun input(arg0: String): CommunicationResult {
        val result: CommunicationResult
        val analyzer = StringAnalyzer(arg0)
        if (analyzer.isMain) {
            mainFunction = this.defaultMainFunction
            mainFunction += analyzer.text
            result = submit()
        } else {
            program += analyzer.text + "\n"
            result = ResultFactory.createAddDefineResult("{\"AddedDefine\":\"" + analyzer.text + "\"}")
        }
        return result
    }

    fun submit(): CommunicationResult {
        return this.compiler.post(program + mainFunction, "")
    }

}
